Improvements for python coding standards (#224)
canonical-web-and-design/practices: coding/python.md

* * *

title: Python code standards
description: Code style standards for Python

* * *

This document outlines the rules for writing Python code across all our Projects.

## Code style and formatting

We use three tools to conform to [PEP 8](https://pep8.org/):

-   [Black](https://black.readthedocs.io/en/stable/). We format our Python code using the option `--line-length 79` followed by the folders that we want to format. E.g. `black --line-length 79 webapp tests`

-   [Flake8](https://flake8.pycqa.org/en/latest/) is a tool to enforce code styling. It will suggest you to remove unsued variables, remove unused imports, enforce more strict whitespace practices and calculate your code complexity. E.g. `flake8 webapp tests`

-   [iSort](https://readthedocs.org/projects/isort/) automatically rearranges your imports in different sections. It separates `import package` from `from pacakage import module` and tides up the spacing between them.

> Project should be configured to block merging on pull requests that do not conform to these standards. You can do this by adding the initials `WIP` to the beginning of your PR title.

Reviewer: Also, we don't at the moment check iSort in CI for any project I'm aware of.

> Project should be configured to block merging on pull requests that do not conform to these standards. You can do this by adding the initials `WIP` to the beginning of your PR title.

Author: Yeah, as discussed, I missed that this was talking about Circle CI/Github actions entirely.

> > If you use VSCode, linting is enabled by default with the Python extension. To use black specifically add `"python.formatting.blackArgs": [ "--line-length", "79"], "python.formatting.provider": "black",` to your `settings.json`. Also you might want to install iSort extension and use Shift + Control (Command on a Mac) and type iSort to automatically order your imports. 

Author: Possibly moving this to a discourse post.

## Testing

Although tests are lacking on many of our existing projects, writing tests is strongly encouraged. 

-   We write our tests in the `tests` folders. Files usually follow the `test_*.py` convention. 
-   After a tests has been written, we use the `python -m unittest discover` to run the tests.

> -   Once the tests are run successfuly, you should check the coverage report by running `coverage report -m` to check that you have passed the minimum required code coverage level to avoid failing Circle CI tests.

Author: For me personally, it is quite annoying to push some code and having to fix a tiny thing (lint or fix tests), that is why I strongly suggest this. But I guess this is optional, I just encourage it for the sake of the developer, not really by any convention or rule.

> -   Once the tests are run successfuly, you should check the coverage report by running `coverage report -m` to check that you have passed the minimum required code coverage level to avoid failing Circle CI tests.

Reviewer: Oh running the linter is great, I agree with that. It's the extra step of generating and checking the coverage report I think might be a little overboard.

Ideally every project should have a `test-python` and `lint-python` script in `package.json` to make this easier.

> Please write tests for any new code if at all possible, and any efforts to improve test coverage on any project would be greatly appreciated.

Reviewer: It would be nice to add a similar line about CI here. Or maybe we move the other CI line to the bottom and make it apply to both tests and formatting.

Author: I'm not so well versed in CI, maybe @jkfran can write up something?

Reviewer: Yeah that's fine, we can add something after you've merged this PR.

## Code style and formatting

> We use [Flake8](https://flake8.pycqa.org/en/latest/) to conform to [PEP 8](https://pep8.org/). It will suggest you to remove unsued variables, remove unused imports, enforce more strict whitespace practices and calculate your code complexity. E.g. `flake8 webapp tests`.

Reviewer: >, remove unused imports, 

I'm not aware of `flake8` removing anything on its own - AFAIK it will only checks, it doesn't format. Or maybe you meant that it will "suggest you to ... remove unused imports", in which case I think the wording could be tidied up.

I was willing to believe you that it complains about unused variables, but in my testing this doesn't seem to be the case.

I am also not aware of it having an capability to calculate code complexity (and I'm not aware of PEP8 talking about code complexity either). It would be great to have something to calculate code complexity though.

I'm don't quite understand what's meant by "enforce strict whitespace practices". I think this could be made clearer.

How about simply:

```suggestion
We use [Flake8](https://flake8.pycqa.org/en/latest/) to check our code for compliance with [PEP 8](https://pep8.org/), e.g. with `flake8 webapp tests`. See [the list of Flake8 violations](https://flake8.pycqa.org/en/latest/user/error-codes.html) for more information.
```

We use [Black](https://black.readthedocs.io/en/stable/) to format our Python code using the option `--line-length 79` followed by the folders that we want to format. E.g. `black --line-length 79 webapp tests`

Optionally, you can use [iSort](https://readthedocs.org/projects/isort/) to automatically rearranges your imports in different sections. It separates `import package` from `from pacakage import module` and tides up the spaces between them. However, this is not enforced by PEP8 guidelines.

> Project should be configured to block merging on pull requests that do not conform to these standards (currently we are using Circle CI but we are migrating to Github Actions).

Reviewer 2: ```suggestion
A project should be configured to block merging on pull requests that do not conform to these standards. (Currently we are using CircleCI, but we are migrating to GitHub Actions.)
```

Reviewer: I think ideally it should be "Projects" rather than "A project". And can we simply remove the line about Circle CI / GitHub Actions as it will go out of date almost immediately?

## Testing

Writing tests is strongly encouraged.

- We write our tests in the `tests` folders. Files usually follow the `test_*.py` (pattern for unittest discovery)[https://docs.python.org/3/library/unittest.html#cmdoption-unittest-discover-p].

> - After a tests has been written, we use the `python3 -m unittest discover --start-directory tests` to run the tests.

Reviewer 2: ```suggestion
- After tests have been written, we use the `python3 -m unittest discover --start-directory tests` to run them.
```

> - Once the tests are run successfuly, you can check the coverage report by running `coverage report -m` to check that you have passed the minimum required code coverage level to avoid failing CI tests.

Reviewer 3: So here there are 2 things:

- the tool to calculate the coverage: https://pypi.org/project/coverage/ This is just a python tool that calculates the coverage of a codebase after running unnitest in our case.
- the tool that takes this coverage and uses this data to block the CI, makes a diff between a previous report etc. : https://pypi.org/project/coverage/

I think we should split this part.

Reviewer: I think you accidentally put the same link twice..

Author: Ok. How about I remove code coverage, as discussed above, we don't want to include any tools that might be outdated in the future.

> - Once the tests are run successfuly, you can check the coverage report by running `coverage report -m` to check that you have passed the minimum required code coverage level to avoid failing CI tests.

Reviewer 2: ```suggestion
- Once the tests are run successfully, you can check the coverage report by running `coverage report -m` to check that you have passed the minimum required code coverage level to avoid failing CI tests.
```
